blog/content/post/2020-11-30-coreboot_for_power.md

```diff
@@ -0,0 +1,154 @@
+---
+title: coreboot port for OpenPOWER - why bother?
+abstract: 'You may have heard by now that we are working on coreboot port for
+          Talos II. OpenPOWER already has, nomen omen, open source firmware, so
+          one may ask why bother? We will try to answer that question.'
+cover: /covers/coreboot-logo.svg
+author: krystian.hebel
+layout: post
+published: true
+date: 2020-11-30
+archives: "2020"
+
+tags:
+  - firmware
+  - coreboot
+  - openpower
+categories:
+  - Firmware
+
+---
+
+You may have heard by now that we are working on coreboot port for [Talos II](https://raptorcs.com/TALOSII/).
+OpenPOWER already has, nomen omen, open source firmware, so one may ask why
+bother? In this blog post we will try to answer that question.
+
+## Short introduction to OpenPOWER boot process and its components
+
+Talos II is a server platform managed by a slightly customised (mainly to handle
+specific dual-CPU boot synchronisation quirks) [OpenBMC fork](https://git.raptorcs.com/git/talos-openbmc/).
+BMC signals the main (host) platform to start booting.
+
+Next comes the [Self-Boot Engine](https://wiki.raptorcs.com/wiki/Self-Boot_Engine).
+This term is actually used both for a chip (which is part of main CPU, it is a
+reduced PowerPC core) and its [firmware](https://git.raptorcs.com/git/talos-sbe/).
+SBE's tasks are initialisation of main CPU cores and loading and staring the
+next component called Hostboot.
```

Comment:
Technically there is another piece in between.  The SBE code loads up the Hostboot Bootloader first.  This piece of code is responsible for loading the Hostboot code from the PNOR, verifying its signature, then jumping into Hostboot code itself.

Note that the Bootloader code is part of the SBE SEEPROM contents.

Reply:
I've tried to keep this list as simple as possible, but now that it gave us some problems I think it is worth describing.

How exactly is SEEPROM updated with newer versions? I know this is done by  the Hostboot, so sometimes there is an additional reboot after flashing, but haven't found the code responsible for it yet (not that I've specifically looked for it). Are there any checks done? We were planning to put coreboot's bootblock in the place of HBBL, which would then parse PNOR and load the rest of stages, but we're not sure if this is feasible without modifications to the SBE code.

Reply:
Probably over i2c? You can read/write the seeprom from the bmc (well, one of each pair in the cpu)
via the i2c sysfs interface; its presented as four logical at24-type eeproms (however the size in the
bmc devicetree is wrong so you have to bind it manually).

Reply:
https://github.com/open-power/hostboot/blob/master/src/usr/sbe/sbe_update.C is the code that drives the SBE update from Hostboot.  

The SBE code doesn't care what it loads so you can freely replace HBBL.  However, you should be aware that the space is extremely limited.  I think we're literally down to having a handful of free bytes left.  If you need more space, that would require changes to the layout of the SBE code.

> You can read/write the seeprom from the bmc 
Note that the SEEPROM is typically not accessible from the BMC, unless you have security disabled with the jumper.

Reply:
Thanks @hanetzer, thanks @dcrowell77, both comments are extremely useful. Back to the drawing board I go :)

Reply:
Would be interesting to turn the sbefw into coreboot as well, with the sbe runtime as its payload.

Reply:
I'm afraid coreboot is too big to fit there. Unless... Are you thinking about starting it on QEMU just for testing?